group/assignment6/assignment6.py

```diff
@@ -75,27 +77,66 @@ def compute_efficient_moves(start_state, end_state):
-                  Each move is represented as a tuple with two indices,
-                  the 1st index is the number of lot from which we move the car,
-                  the 2nd index is the number of lot to which we move the car
-    :returns: end state, list of car numbers.
+    :param end_state: order of cars after rearrangement
+    :param constraints: map from the parking lot to a tuple of the allowed cars
+    :yields: move steps. Each move is represented as a tuple with two indeces,
+             the 1st index is the number of lot from which we move the car,
+             the 2nd index is the number of lot to which we move the car
     """
-    parking = Parking(start_state.copy())
-    for move in moves:
-        move_from, move_to = move
-        parking.move_to_empty_lot(parking.get_car(move_from))
-    return parking.get_state()
+    path_finder = PathFinder(tuple(start_state), tuple(end_state), constraints)
+    paths = path_finder.find_all_paths()
+    path = path_finder.decode_path(next(paths))
+    for i in range(1, len(path)):
+        yield compute_move(path[i - 1], path[i])
 
 
+def compute_all_moves(start_state, end_state):
+    """
+    Computes all possible ways to rearrange the cars from start_state to end_state.
+
+    :param start_state: order of cars in the start of the rearrangement
+    :param end_state: order of cars after rearrangement
+    :yields: a list of move steps. Each move is represented as a tuple with two indeces,
+             the 1st index is the number of lot from which we move the car,
+             the 2nd index is the number of lot to which we move the car
+    """
+    parking_size = len(start_state)
+    path_finder = PathFinder(tuple(start_state), tuple(end_state), constraints={}) 
+    paths = path_finder.find_all_paths()
+    for path in paths:
+        decoded_path = path_finder.decode_path(path)
+        moves_sequence = list()
+        for i in range(1, len(decoded_path)):
+            moves_sequence.append(compute_move(decoded_path[i - 1], decoded_path[i]))
+        yield moves_sequence
+    
+
 if __name__ == '__main__':
     start_state = [1, 2, 0, 3] 
     end_state = [3, 1, 2, 0]
+
+    print("Computing a moves sequence for the main problem...")
     moves = list(compute_moves(start_state, end_state))
     print(moves)
-    print(apply_moves(start_state, moves))
+    
+    print("Computing the shortest moves sequence...")
+    moves = list(compute_efficient_moves(start_state, end_state))
+    print(moves)
+
+    print("Computing a moves sequence under given constraints...")
+    constraints = {0: (1, 3), 1: (0, 1, 2, 3), 2: (0, 1, 2, 3), 3: (0, 1, 2, 3)}
+    moves = list(compute_moves_with_constraints(start_state, end_state, constraints))
+    print(moves)
+
+    print("Computing all moves sequences between two parking states...")
+    moves_sequences = list(compute_all_moves(start_state, end_state))
+    n_sequences = len(moves_sequences)
```

Comment:
We shouldn't create a list of our generator because it will be computed and will take a long time, so we need to pass the number of requirement sequences as a parameter and then return first sequences o to avoid long (infinite) running of our function by computing all possible sequences